streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -761,7 +761,7 @@ private void runLoop() {
                     "Will close out all assigned tasks and rejoin the consumer group.");
 
                 taskManager.handleLostAll();
```

Comment:
Do we still need this here? When we call unsubscribe later, it'll invoke:
```
            if (generation() != Generation.NO_GENERATION) {
                e = invokePartitionsRevoked(droppedPartitions);
            } else {
                e = invokePartitionsLost(droppedPartitions);
            }
```
It seems like it might be better to just let the consumer manage the task lifecycle, since we may not have lost _all_ the tasks.

WDYT about removing this line and just setting the `needRebalance` flag?

Reply:
Make sense to me, either way we should be able to hit `handleLostAll`, just one question would be, could we guarantee that the tasks are always matching the assigned partitions? If consumer doesn't own any partition but we still have tasks not cleaned up, this could block us from triggering the callback.

Reply:
I'm not 100% sure, but I'd be surprised if we could ever have a task for a partition we don't have assigned. If you're concerned about it, we can just disregard this topic for now.

Reply:
@vvcephei @abbccdda there happen to have a discussion that validates this :) https://github.com/apache/kafka/pull/8087/files#r382735615

Reply:
I agree we should remove the direct `handleLostAll` call here and leave it to `enforceRebalance` to appropriately handle the `TaskMigratedException` scenario

Reply:
IMO we should never call the `handleXXX` methods from outside the rebalance callbacks

Reply:
Actually, due to a race condition with the heartbeat thread it seems we are not safe to assume `unsubscribe` will invoke `onPartitionsLost` in which case we would end up throwing `TaskMigratedException` due to invoking `onPartitionsRevoked`. I think we should fix this more soundly in trunk at the consumer level, but for 2.5 the best way/only way to handle it from Streams is to catch this exception and then invoke `handleLostAll` directly from the Streamthread. See discussion